ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java

```diff
@@ -229,20 +219,79 @@ public void run() {
   }
 
   /**
-   * Replicas from the given replicas
-   * @param replicasToReplicate list of {@link RemoteReplicaInfo} by data node
+   * Remove {@link RemoteReplicaInfo} from current {@link ReplicaThread}.
+   * @param remoteReplicaInfo {@link RemoteReplicaInfo} to remove.
+   */
+  void removeRemoteReplicaInfo(RemoteReplicaInfo remoteReplicaInfo) {
+    lock.lock();
+    try {
+      DataNodeId dataNodeId = remoteReplicaInfo.getReplicaId().getDataNodeId();
+      Set<RemoteReplicaInfo> remoteReplicaInfos = replicasToReplicateGroupedByNode.get(dataNodeId);
+      if (remoteReplicaInfos != null) {
+        if (!remoteReplicaInfos.remove(remoteReplicaInfo)) {
+          replicationMetrics.remoteReplicaInfoRemoveError.inc();
+          logger.error("ReplicaThread: {}, RemoteReplicaInfo {} not found.", threadName, remoteReplicaInfo);
+        }
+      } else {
+        replicationMetrics.remoteReplicaInfoRemoveError.inc();
+        logger.error("ReplicaThread: {}, RemoteReplicaInfos Set is not created for DataNode {}, RemoteReplicaInfo: {}.",
+            threadName, dataNodeId, remoteReplicaInfo);
+      }
+    } finally {
+      lock.unlock();
+    }
+    logger.trace("RemoteReplicaInfo {} is removed from ReplicaThread {}.", remoteReplicaInfo, threadName);
+  }
+
+  /**
+   * Add a {@link RemoteReplicaInfo} to current {@link ReplicaThread}.
+   * @param remoteReplicaInfo {@link RemoteReplicaInfo} to add.
+   */
+  public void addRemoteReplicaInfo(RemoteReplicaInfo remoteReplicaInfo) {
+    lock.lock();
+    try {
```

Comment:
I don't see the necessity to add `try` block here.

Reply:
That's a safe guard. @lightningrob suggests add try{} finally{} for all lock operation. (In case some runtime exception: NPE)

Reply:
+1, if possible, unlock calls should go in the finally block so that the lock is guaranteed to be released at some point

Reply:
I see. Good to know. Thanks